File: clinglib/src/main/java/me/shetj/cling/manager/DeviceManager.java
package me.shetj.cling.manager;


import android.content.Context;
import android.util.Log;

import java.util.Collection;

import me.shetj.cling.control.SubscriptionControl;
import me.shetj.cling.entity.ClingDevice;
import me.shetj.cling.entity.ClingDeviceList;
import me.shetj.cling.entity.IDevice;
import me.shetj.cling.util.Utils;


public class DeviceManager <T> implements IDeviceManager <T> {
    private static final String TAG = DeviceManager.class.getSimpleName();
    /**
     * 已选中的设备, 它也是 ClingDeviceList 中的一员
     */
    private ClingDevice mSelectedDevice;
    private SubscriptionControl mSubscriptionControl;

    public DeviceManager() {
        mSubscriptionControl = new SubscriptionControl();
    }

    @Override
    public IDevice getSelectedDevice() {
        return mSelectedDevice;
    }

    @Override
    public void setSelectedDevice(IDevice selectedDevice) {

        Log.i(TAG, "Change selected device.");
        mSelectedDevice = (ClingDevice) selectedDevice;
        // 重置选中状态
        Collection<ClingDevice> clingDeviceList = ClingDeviceList.getInstance().getClingDeviceList();
        if (Utils.isNotNull(clingDeviceList)){
            for (ClingDevice device : clingDeviceList){
                device.setSelected(false);
            }
        }
        // 设置选中状态
        mSelectedDevice.setSelected(true);
    }

    @Override
    public void cleanSelectedDevice() {
        if (Utils.isNull(mSelectedDevice))
            return;

        mSelectedDevice.setSelected(false);
    }

    @Override
    public void registerAVTransport(Context context) {
        if (Utils.isNull(mSelectedDevice))
            return;

        mSubscriptionControl.registerAVTransport(mSelectedDevice, context);
    }

    @Override
    public void registerRenderingControl(Context context) {
        if (Utils.isNull(mSelectedDevice))
            return;

        mSubscriptionControl.registerRenderingControl(mSelectedDevice, context);
    }

    @Override
    public void destroy() {
        if (Utils.isNotNull(mSubscriptionControl)){
            mSubscriptionControl.destroy();
        }
    }
}

File: clinglib/src/main/java/me/shetj/cling/manager/DeviceManager.kt
package me.shetj.cling.manager

import android.content.Context
import me.shetj.cling.control.SubscriptionControl
import me.shetj.cling.entity.ClingDevice
import me.shetj.cling.entity.ClingDeviceList.getClingDeviceList
import me.shetj.cling.util.Utils.isNotNull
import me.shetj.cling.util.Utils.isNull

internal class  DeviceManager  : IDeviceManager  {

    private var mSelectedDevice: ClingDevice? = null
    private var mSubscriptionControl: SubscriptionControl  = SubscriptionControl()


    override fun cleanSelectedDevice() {
        if (isNull(mSelectedDevice)) return

        mSelectedDevice!!.isSelected = false
    }

    override fun registerAVTransport(context: Context) {
        if (isNull(mSelectedDevice)) {
            return
        }
        mSubscriptionControl.registerAVTransport(mSelectedDevice!!  , context)
    }

    override fun registerRenderingControl(context: Context) {
        if (isNull(mSelectedDevice)) {
            return
        }
        mSubscriptionControl.registerRenderingControl(mSelectedDevice!!, context)
    }

    override fun destroy() {
        if (isNotNull(mSubscriptionControl)) {
            mSubscriptionControl.destroy()
        }
    }


    override fun setSelectedDevice(device: ClingDevice) {
        mSelectedDevice = device
        // 重置选中状态
        val clingDeviceList: Collection<ClingDevice> =
            getClingDeviceList()
        if (isNotNull(clingDeviceList)) {
            clingDeviceList.forEach {
                it.isSelected = false
            }
        }
        // 设置选中状态
        mSelectedDevice?.isSelected = true
    }

    override fun getSelectedDevice(): ClingDevice? {
        return mSelectedDevice
    }

}